http/agent.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright © The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	defaultPostContentType = "application/octet-stream"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+// Agent is an http agent
+type Agent struct {
+	options *AgentOptions
+	AgentImplementation
+}
+
+// AgentImplementation is the actual implementation of the http calls
+//counterfeiter:generate . AgentImplementation
+type AgentImplementation interface {
+	SendPostRequest(string, []byte, *AgentOptions) (*http.Response, error)
+	SendGetRequest(string, *AgentOptions) (*http.Response, error)
+}
+
+type defaultAgentImplementation struct{}
+
+// AgentOptions has the configurable bits of the agent
+type AgentOptions struct {
```

Comment:
We could make the options type private. This way the only API surface for changing the behavior would be the `With…` methods.

Reply:
:thinking:  Mmh I tried doing this at first, but then the problem is that I need the type to be exported in the fakes:
```golang
func (fake *FakeAgentImplementation) SendGetRequest(
    arg1 string, arg2 *http.AgentOptions,
) (*httpa.Response, error) 
```

Reply:
OK, I've reworked the implementation functions to not require the options. Now the options type is unexported.